iosApp/iosApp/ViewModels/NearbyViewModel.swift

```diff
@@ -116,14 +116,40 @@ class NearbyViewModel: ObservableObject {
             )
             pushNavEntry(.stopDetails(stopId: stopId, stopFilter: stopFilter, tripFilter: tripFilter))
         } else if case let .legacyStopDetails(targetStop, _) = entry,
-                  case let .legacyStopDetails(currentStop, _) = navigationStack.last,
-                  targetStop == currentStop {
+                  case let .legacyStopDetails(lastStop, _) = navigationStack.last,
+                  targetStop == lastStop {
             _ = navigationStack.popLast()
             navigationStack.append(entry)
-        } else if case let .stopDetails(stopId: targetStop, stopFilter: _, tripFilter: _) = entry,
-                  case let .stopDetails(stopId: currentStop, stopFilter: _, tripFilter: _) = navigationStack.last,
-                  targetStop == currentStop {
+        } else if
+            case let .stopDetails(
+                stopId: targetStop,
+                stopFilter: newFilter,
+                tripFilter: _
+            ) = entry,
+            case let .stopDetails(
+                stopId: lastStop,
+                stopFilter: lastFilter,
+                tripFilter: _
+            ) = navigationStack.last,
+            lastFilter != nil,
+            targetStop == lastStop {
+            // When the stop filter changes, we want a new entry to be added (i.e. no pop) only when
+            // you're on the unfiltered (lastFilter == nil) page, but if there is already a filter,
+            // the entry with the old filter should be popped and replaced with the new value.
             _ = navigationStack.popLast()
+
+            // If the new filter is nil and there is already a nil filter in the stack for the same stop ID,
+            // we don't want a duplicate unfiltered entry, so skip appending a new one
```

Comment:
Should this condition really be inside the `lastFilter != nil`? It isn't in the other copy of this logic. (It would probably be worth refactoring this logic out somewhere else even if the result is a bit ugly at each call site.)

Reply:
Good call, how's that? I didn't include the pop check in the function because I didn't want to give the extension function side effects.